docs/architecture/adr-102-rpc-companion.md

```diff
@@ -0,0 +1,358 @@
+problem.
+
+One way is to support a mechanism to migrate the old data to the new data structures. This would require additional logic for the migration process
+that would need to be run before ingesting data from an upgraded full node that contains the new data structures. But this approach might also
+cause some issues, for example, if the new data structure has a new field that there's no corresponding value in the "old" data structure, probably
+the value would need to be set to `[null]` but this can have unintended consequences. It's important to ensure that in the future if this solution
+is largely adopted and used, then a change request that affects the schema as outlined above should also contain the details needs to ensure how
+PostgreSQL should handle this change.
+
+
+Another potential solution for this scenario is to find a way in the database that can support "versioning" of data structures. For example, let's
+assume there's a `Block` structure, let's call it `v1`. If in the future there's a need to modify this structure that is not compatible with the
+previous data structure, then the database would support a `v2` schema for `ResultBlock` and an `index` table could determine the criteria on which
+data structure should be used for inserting or querying data.
+
+There's also a possibility that the data structure could be stored in parallel with the normalized data but in one blob, e.g. in a `jsonb` field
+that contains all the information needed to return the client, this could be an optimized way to serve the data but could add to the storage
+requirements. This could improve the performance for reading data from the database. An alternative approach to this problem that can offer
+similar results would be to use a "caching" layer.
+
+
+### [RPC server instance](#rpc-instance)
+
+The **RPC server instance** is a node that runs the RPC API process for the data companion. This server instance provides an RPC API (`/v1`) with
+the same JSONRPC methods of the full node JSONRPC endpoint. The RPC Companion service will expose the same JSONRPC methods and will accept the same request types and
+return wire compatible responses (should match the same response as the equivalent full node JSONRPC endpoint).
+
+The **RPC server instance**, when serving a particular request, retrieves the required data from the database in order to
+fulfill the request. The data should be serialized in a way that makes it wire compatible with the CometBFT JSONRPC endpoint.
+
+The RPC service endpoint should be exposed through an external load-balancer service such as Cloudflare or AWS ELB, or
+a server running its own load balancer mechanism (e.g. nginx).
+
+The RPC clients should make requests to the **RPC Companion** server instances through this load balancer.
+
+The **RPC server instance** will also implement logic for the proxied requests to the full node. It should properly handle the proxy requests and responses from the full node.
```

Comment:
Let's also mention that the `RPC server instance` should cache responses from the full node to avoid requesting the same data over and over again. Wdyt?

Reply:
added text for caching, thanks